quantum/audio/audio_avr_pwm.c

```diff
@@ -0,0 +1,307 @@
+        #define AUDIO2_COMxy0 COM3A0
+        #define AUDIO2_COMxy1 COM3A1
+        #define AUDIO2_OCIExy OCIE3A
+        #define AUDIO2_OCRxy OCR3A
+        #define AUDIO2_PORTxy PORTC6
+        #define AUDIO2_TIMERx_COMPy_vect TIMER3_COMPA_vect
+    #elif defined(C5_AUDIO)
+        #define AUDIO2_COMxy0 COM3B0
+        #define AUDIO2_COMxy1 COM3B1
+        #define AUDIO2_OCIExy OCIE3B
+        #define AUDIO2_OCRxy OCR3B
+        #define AUDIO2_PORTxy PORTC5
+        #define AUDIO2_TIMERx_COMPy_vect TIMER3_COMPB_vect
+    #elif defined(C4_AUDIO)
+        #define AUDIO2_COMxy0 COM3C0
+        #define AUDIO2_COMxy1 COM3C1
+        #define AUDIO2_OCIExy OCIE3C
+        #define AUDIO2_OCRxy OCR3C
+        #define AUDIO2_PORTxy PORTC4
+        #define AUDIO2_TIMERx_COMPy_vect TIMER3_COMPC_vect
+    #endif
+#endif
+
+
+// C6 seems to be the assumed default by many existing keyboard - but sill warn the user
+#if !defined(AUDIO2_PIN_SET) && !defined(AUDIO1_PIN_SET)
+#pragma message "audio feature enabled, but no pin selected - see docs/feature_audio for defines to set"
+#endif
+// -----------------------------------------------------------------------------
+
+
+float place = 0;
+
+#ifdef AUDIO1_PIN_SET
+static float channel_1_frequency = 0.0f;
+void channel_1_set_frequency(float freq) {
+
+    if (freq < 0.0f)
+        freq = 0.0f;
+
+    if (freq == channel_1_frequency)
+        return;
+
+    channel_1_frequency = freq;
+
+    // set pwm period
+    AUDIO1_ICRx = (uint16_t)(((float)F_CPU) / (freq * CPU_PRESCALER));
+    // and duty cycle
+    AUDIO1_OCRxy = (uint16_t)((((float)F_CPU) / (freq * CPU_PRESCALER)) * note_timbre);
+}
+
+float channel_1_get_frequency(void) {
+    return channel_1_frequency;
+}
+
+void channel_1_start(void){
+    // enable timer-counter ISR
+    AUDIO1_TIMSKx |= _BV(AUDIO1_OCIExy);
+    // enable timer-counter output
+    AUDIO1_TCCRxA |= _BV(AUDIO1_COMxy1);
+}
+
+void channel_1_stop(void){
+    // disable timer-counter ISR
+    AUDIO1_TIMSKx &= ~_BV(AUDIO1_OCIExy);
+    // disable timer-counter output
+    AUDIO1_TCCRxA &= ~(_BV(AUDIO1_COMxy1) | _BV(AUDIO1_COMxy0));
+}
+#endif
+
+#ifdef AUDIO2_PIN_SET
+static float channel_2_frequency = 0.0f;
+void channel_2_set_frequency(float freq) {
+
+    if (freq < 0.0f)
+        freq = 0.0f;
+
+    if (freq == channel_2_frequency)
+        return;
+
+    channel_2_frequency = freq;
+
+    AUDIO2_ICRx = (uint16_t)(((float)F_CPU) / (freq * CPU_PRESCALER));
+    AUDIO2_OCRxy = (uint16_t)((((float)F_CPU) / (freq * CPU_PRESCALER)) * note_timbre);
+}
+
+void channel_2_start(void){
+    AUDIO2_TIMSKx |= _BV(AUDIO2_OCIExy);
+    AUDIO2_TCCRxA |= _BV(AUDIO2_COMxy1);
+}
+
+void channel_2_stop(void){
+    AUDIO2_TIMSKx &= ~_BV(AUDIO2_OCIExy);
+    AUDIO2_TCCRxA &= ~(_BV(AUDIO2_COMxy1) | _BV(AUDIO2_COMxy0));
+}
+#endif
+
+
+void audio_initialize_hardware() {
+
+#ifdef AUDIO1_PIN_SET
+    channel_1_stop();
+    // Set audio ports as output
+    AUDIO1_DDRx |= _BV(AUDIO1_PORTxy);
```

Comment:
I suspect we can also do away with these two defines by defining eg. `AUDIO1_PIN B5` in the relevant places, and calling [`setPinOutput(AUDIO1_PIN)`](https://github.com/qmk/qmk_firmware/blob/7f23ad72c5c736b58bab16995e7b0a599268b56f/quantum/quantum.h#L161) here instead. It may require including quantum.h though.

Reply:
good point - hat a look at quantum.h and the define: their purpose seems to be platform-agnostic pin-manipulation - which won't matter here since this file is AVR specific (hence the filename :-D).

Reply:
It's more self-documenting though, which is what you were after here ;)

As an example, incoming keyboards and keymaps are encouraged to use these defines as they're much easier to read. I believe this also applies to core code.

Reply:
ok - leaving the defines as they are :-)